subsys/bluetooth/audio/ascs.c

```diff
@@ -3082,20 +3083,98 @@ static ssize_t ascs_cp_write(struct bt_conn *conn,
 #define BT_ASCS_ASE_SRC_DEFINE(_n, ...) BT_ASCS_ASE_DEFINE(BT_UUID_ASCS_ASE_SRC, (_n) + 1 + \
 							   CONFIG_BT_ASCS_ASE_SNK_COUNT)
 
-BT_GATT_SERVICE_DEFINE(ascs_svc,
-	BT_GATT_PRIMARY_SERVICE(BT_UUID_ASCS),
-	BT_AUDIO_CHRC(BT_UUID_ASCS_ASE_CP,
-		      BT_GATT_CHRC_WRITE | BT_GATT_CHRC_WRITE_WITHOUT_RESP | BT_GATT_CHRC_NOTIFY,
-		      BT_GATT_PERM_WRITE_ENCRYPT | BT_GATT_PERM_PREPARE_WRITE,
-		      NULL, ascs_cp_write, NULL),
-	BT_AUDIO_CCC(ascs_cp_cfg_changed),
+#define BT_ASCS_CHR_ASE_CONTROL_POINT \
+	BT_AUDIO_CHRC(BT_UUID_ASCS_ASE_CP, \
+		      BT_GATT_CHRC_WRITE | BT_GATT_CHRC_WRITE_WITHOUT_RESP | BT_GATT_CHRC_NOTIFY, \
+		      BT_GATT_PERM_WRITE_ENCRYPT | BT_GATT_PERM_PREPARE_WRITE, \
+		      NULL, ascs_cp_write, NULL), \
+	BT_AUDIO_CCC(ascs_cp_cfg_changed)
+
 #if CONFIG_BT_ASCS_ASE_SNK_COUNT > 0
+#define BT_ASCS_ASE_SINKS \
 	LISTIFY(CONFIG_BT_ASCS_ASE_SNK_COUNT, BT_ASCS_ASE_SNK_DEFINE, (,)),
+#else
+#define BT_ASCS_ASE_SINKS
 #endif /* CONFIG_BT_ASCS_ASE_SNK_COUNT > 0 */
+
 #if CONFIG_BT_ASCS_ASE_SRC_COUNT > 0
+#define BT_ASCS_ASE_SOURCES \
 	LISTIFY(CONFIG_BT_ASCS_ASE_SRC_COUNT, BT_ASCS_ASE_SRC_DEFINE, (,)),
+#else
+#define BT_ASCS_ASE_SOURCES
 #endif /* CONFIG_BT_ASCS_ASE_SRC_COUNT > 0 */
-);
+
+#define BT_ASCS_SERVICE_DEFINITION() { \
+	BT_GATT_PRIMARY_SERVICE(BT_UUID_ASCS), \
+	BT_AUDIO_CHRC(BT_UUID_ASCS_ASE_CP, \
+		      BT_GATT_CHRC_WRITE | BT_GATT_CHRC_WRITE_WITHOUT_RESP | BT_GATT_CHRC_NOTIFY, \
+		      BT_GATT_PERM_WRITE_ENCRYPT | BT_GATT_PERM_PREPARE_WRITE, \
+		      NULL, ascs_cp_write, NULL), \
+	BT_AUDIO_CCC(ascs_cp_cfg_changed), \
+	BT_ASCS_ASE_SINKS \
+	BT_ASCS_ASE_SOURCES \
+	}
+
+#define ASCS_ASE_CHAR_ATTR_COUNT 3 /* declaration + value + cccd */
+
+static struct bt_gatt_attr ascs_attrs[] = BT_ASCS_SERVICE_DEFINITION();
+static struct bt_gatt_service ascs_svc = (struct bt_gatt_service)BT_GATT_SERVICE(ascs_attrs);
+
+static void remove_ase_char(uint8_t snk_cnt, uint8_t src_cnt)
+{
+	uint8_t snk_ases_to_rem = CONFIG_BT_ASCS_ASE_SNK_COUNT - snk_cnt;
+	uint8_t src_ases_to_rem = CONFIG_BT_ASCS_ASE_SRC_COUNT - src_cnt;
+	size_t attrs_to_rem;
```

Comment:
I assume it's easier to remove characteristics, than to add them?

Reply:
That was my experience yes :)
Less important, but subjectively, in the end it also made more sense to me, if considering the KConfig configuration as a default setting.

Reply:
Indeed. I did the same for CSIP now that I remember: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/audio/csip_set_member.c#L690-L732